File: api/app/project_config.py
"""Project config schema + inherit/override resolution (pure helpers).

A project owns a ``defaults`` document with four groups (scope/capture/denylist/
analysis). A session resolves its effective config once, at creation, from the
project defaults plus a sparse set of per-field overrides: null = inherit, set =
replace (per leaf; list values are replaced, never unioned). These pure helpers
are the single source of truth for the shape and the merge. No DB, stdlib only."""
import copy
from typing import Any

_OUT_OF_SCOPE_MODES = {"tag", "mute", "exclude"}

SYSTEM_DEFAULTS: dict[str, Any] = {
    "scope": {"rootDomains": [], "includeSubdomains": True},
    "capture": {"outOfScopeMode": "tag", "maxAssetMb": 10},
    "denylist": {"rules": [], "useDefaultProfile": True},
    "analysis": {"analyzeOnUpload": False, "captureSourceMaps": True},
}

# Every leaf a project owns and a session may override, grouped by section.
CONFIG_SCHEMA: dict[str, tuple[str, ...]] = {
    "scope": ("rootDomains", "includeSubdomains"),
    "capture": ("outOfScopeMode", "maxAssetMb"),
    "denylist": ("rules", "useDefaultProfile"),
    "analysis": ("analyzeOnUpload", "captureSourceMaps"),
}


def system_defaults() -> dict[str, Any]:
    return copy.deepcopy(SYSTEM_DEFAULTS)


def deep_merge(base: dict, patch: dict) -> dict:
    """Recursively merge ``patch`` into a copy of ``base``; patch leaves win.
    Used to apply a partial project-defaults update over the stored document."""
    out = copy.deepcopy(base)
    for key, value in (patch or {}).items():
        if isinstance(value, dict) and isinstance(out.get(key), dict):
            out[key] = deep_merge(out[key], value)
        else:
            out[key] = copy.deepcopy(value)
    return out


def validate_config(doc: dict, *, partial: bool = False) -> dict:
    """Validate a config document against the schema. With ``partial=True`` only
    the sections that are present are checked (used for per-session captureConfig
    and partial project PATCH). Raises ValueError with a human-readable message."""
    if not isinstance(doc, dict):
        raise ValueError("config must be an object")
    for section in CONFIG_SCHEMA:
        if section not in doc:
            if partial:
                continue
            raise ValueError(f"missing config section: {section}")
        if not isinstance(doc[section], dict):
            raise ValueError(f"config section {section} must be an object")

    if "scope" in doc:
        scope = doc["scope"]
        if "rootDomains" in scope and not isinstance(scope["rootDomains"], list):
            raise ValueError("scope.rootDomains must be a list")
        if "includeSubdomains" in scope and not isinstance(scope["includeSubdomains"], bool):
            raise ValueError("scope.includeSubdomains must be a boolean")
    if "capture" in doc:
        capture = doc["capture"]
        if "outOfScopeMode" in capture and capture["outOfScopeMode"] not in _OUT_OF_SCOPE_MODES:
            raise ValueError("capture.outOfScopeMode must be one of tag|mute|exclude")
        if "maxAssetMb" in capture:
            mb = capture["maxAssetMb"]
            if isinstance(mb, bool) or not isinstance(mb, (int, float)) or mb <= 0 or mb > 10:
                raise ValueError("capture.maxAssetMb must be a number in (0, 10]")
    if "denylist" in doc:
        denylist = doc["denylist"]
        if "rules" in denylist:
            if not isinstance(denylist["rules"], list):
                raise ValueError("denylist.rules must be a list")
            for rule in denylist["rules"]:
                if not isinstance(rule, dict) or "pattern" not in rule:
                    raise ValueError("each denylist rule must be an object with a 'pattern'")
        if "useDefaultProfile" in denylist and not isinstance(denylist["useDefaultProfile"], bool):
            raise ValueError("denylist.useDefaultProfile must be a boolean")
    if "analysis" in doc:
        analysis = doc["analysis"]
        for key in ("analyzeOnUpload", "captureSourceMaps"):
            if key in analysis and not isinstance(analysis[key], bool):
                raise ValueError(f"analysis.{key} must be a boolean")
    return doc


def resolve_effective_config(defaults: dict, overrides: dict | None) -> tuple[dict, list[str]]:
    """Resolve a session's effective config from project defaults + sparse overrides.
    Per leaf: use the override if present, else inherit. Returns (effective, override_keys)
    where override_keys is the sorted dotted paths that were overridden."""
    overrides = overrides or {}
    effective = copy.deepcopy(defaults)
    override_keys: list[str] = []
    for section, keys in CONFIG_SCHEMA.items():
        section_override = overrides.get(section)
        if not isinstance(section_override, dict):
            continue
        for key in keys:
            if key in section_override:
                effective.setdefault(section, {})[key] = copy.deepcopy(section_override[key])
                override_keys.append(f"{section}.{key}")
    return effective, sorted(override_keys)


def split_effective(effective: dict) -> tuple[dict, dict]:
    """Split a resolved config into the scope part (stored in session columns) and
    the capture_config part (capture/denylist/analysis, stored as one JSON column)."""
    scope_section = effective.get("scope") or {}
    scope = {
        "rootDomains": list(scope_section.get("rootDomains") or []),
        "includeSubdomains": bool(scope_section.get("includeSubdomains", True)),
    }
    capture_config = {
        section: copy.deepcopy(effective.get(section) or {})
        for section in ("capture", "denylist", "analysis")
    }
    return scope, capture_config

File: js-security-extractor/chrome-extension/modules/project-config.js
// Client mirror of api/app/project_config.py. Pure, stdlib-only — no chrome/fetch/DOM — so it
// is imported by the ES-module service worker (background.js), bundled into the popup by
// esbuild, and unit-tested directly (node tests/test_project_config.mjs).
//
// A project owns a `defaults` document with four groups (scope/capture/denylist/analysis). A
// session resolves its effective config once — resolveEffectiveConfig(defaults, overrides):
// null/absent = inherit, set = replace (per leaf; lists replace, never union). This file is
// the single source of truth on the client for the shape, the merge, AND the mapping to/from
// the extension's flat chrome.storage settings bag (the capture gate).

const OUT_OF_SCOPE_MODES = new Set(['tag', 'mute', 'exclude']);
const clone = (value) => JSON.parse(JSON.stringify(value === undefined ? null : value));

export const SYSTEM_DEFAULTS = {
  scope: { rootDomains: [], includeSubdomains: true },
  capture: { outOfScopeMode: 'tag', maxAssetMb: 10 },
  denylist: { rules: [], useDefaultProfile: true },
  analysis: { analyzeOnUpload: false, captureSourceMaps: true },
};

// Every leaf a project owns and a session may override, grouped by section.
export const CONFIG_SCHEMA = {
  scope: ['rootDomains', 'includeSubdomains'],
  capture: ['outOfScopeMode', 'maxAssetMb'],
  denylist: ['rules', 'useDefaultProfile'],
  analysis: ['analyzeOnUpload', 'captureSourceMaps'],
};

// The only bridge between the grouped schema and the extension's FLAT storage keys.
// [section, leaf, storageKey]
const SETTINGS_MAP = [
  ['scope', 'rootDomains', 'domainScopes'],
  ['scope', 'includeSubdomains', 'includeSubdomains'],
  ['capture', 'outOfScopeMode', 'outOfScopeMode'],
  ['capture', 'maxAssetMb', 'maxAssetMb'],
  ['denylist', 'rules', 'denyRules'],
  ['denylist', 'useDefaultProfile', 'denyDefaultProfile'],
  ['analysis', 'analyzeOnUpload', 'performAnalysisOnUpload'],
  ['analysis', 'captureSourceMaps', 'captureSourceMaps'],
];

export function systemDefaults() {
  return clone(SYSTEM_DEFAULTS);
}

export function deepMerge(base, patch) {
  const out = clone(base);
  const isObj = (v) => v && typeof v === 'object' && !Array.isArray(v);
  for (const [key, value] of Object.entries(patch || {})) {
    if (isObj(value) && isObj(out[key])) out[key] = deepMerge(out[key], value);
    else out[key] = clone(value);
  }
  return out;
}

export function validateConfig(doc, { partial = false } = {}) {
  if (!doc || typeof doc !== 'object' || Array.isArray(doc)) {
    throw new Error('config must be an object');
  }
  for (const section of Object.keys(CONFIG_SCHEMA)) {
    if (!(section in doc)) {
      if (partial) continue;
      throw new Error(`missing config section: ${section}`);
    }
    const s = doc[section];
    if (!s || typeof s !== 'object' || Array.isArray(s)) {
      throw new Error(`config section ${section} must be an object`);
    }
  }
  if ('scope' in doc) {
    const scope = doc.scope;
    if ('rootDomains' in scope && !Array.isArray(scope.rootDomains)) {
      throw new Error('scope.rootDomains must be a list');
    }
    if ('includeSubdomains' in scope && typeof scope.includeSubdomains !== 'boolean') {
      throw new Error('scope.includeSubdomains must be a boolean');
    }
  }
  if ('capture' in doc) {
    const capture = doc.capture;
    if ('outOfScopeMode' in capture && !OUT_OF_SCOPE_MODES.has(capture.outOfScopeMode)) {
      throw new Error('capture.outOfScopeMode must be one of tag|mute|exclude');
    }
    if ('maxAssetMb' in capture) {
      const mb = capture.maxAssetMb;
      if (typeof mb !== 'number' || Number.isNaN(mb) || mb <= 0 || mb > 10) {
        throw new Error('capture.maxAssetMb must be a number in (0, 10]');
      }
    }
  }
  if ('denylist' in doc) {
    const denylist = doc.denylist;
    if ('rules' in denylist) {
      if (!Array.isArray(denylist.rules)) throw new Error('denylist.rules must be a list');
      for (const rule of denylist.rules) {
        if (!rule || typeof rule !== 'object' || Array.isArray(rule) || !('pattern' in rule)) {
          throw new Error("each denylist rule must be an object with a 'pattern'");
        }
      }
    }
    if ('useDefaultProfile' in denylist && typeof denylist.useDefaultProfile !== 'boolean') {
      throw new Error('denylist.useDefaultProfile must be a boolean');
    }
  }
  if ('analysis' in doc) {
    const analysis = doc.analysis;
    for (const key of ['analyzeOnUpload', 'captureSourceMaps']) {
      if (key in analysis && typeof analysis[key] !== 'boolean') {
        throw new Error(`analysis.${key} must be a boolean`);
      }
    }
  }
  return doc;
}

export function resolveEffectiveConfig(defaults, overrides) {
  const ov = overrides || {};
  const effective = clone(defaults);
  const overrideKeys = [];
  for (const [section, keys] of Object.entries(CONFIG_SCHEMA)) {
    const sectionOverride = ov[section];
    if (!sectionOverride || typeof sectionOverride !== 'object' || Array.isArray(sectionOverride)) continue;
    for (const key of keys) {
      // null/absent = inherit (mirror api/app/project_config.py: `section_override.get(key) is not None`).
      if (sectionOverride[key] != null) {
        if (!effective[section] || typeof effective[section] !== 'object') effective[section] = {};
        effective[section][key] = clone(sectionOverride[key]);
        overrideKeys.push(`${section}.${key}`);
      }
    }
  }
  overrideKeys.sort();
  return { effective, overrideKeys };
}

export function splitEffective(effective) {
  const scopeSection = (effective && effective.scope) || {};
  const scope = {
    rootDomains: Array.isArray(scopeSection.rootDomains) ? [...scopeSection.rootDomains] : [],
    includeSubdomains: scopeSection.includeSubdomains !== false,
  };
  const captureConfig = {};
  for (const section of ['capture', 'denylist', 'analysis']) {
    captureConfig[section] = clone((effective && effective[section]) || {});
  }
  return { scope, captureConfig };
}

// Flat live settings -> grouped effective config. Used for Standalone: the "defaults" a
// project-less session resolves against are the extension's current global settings. Fallbacks
// mirror background.js loadSettings (booleans default true via !== false; mode defaults 'tag').
export function configFromSettings(settings) {
  const s = settings || {};
  return {
    scope: {
      rootDomains: Array.isArray(s.domainScopes) ? [...s.domainScopes] : [],
      includeSubdomains: s.includeSubdomains !== false,
    },
    capture: {
      outOfScopeMode: OUT_OF_SCOPE_MODES.has(s.outOfScopeMode) ? s.outOfScopeMode : 'tag',
      maxAssetMb: (typeof s.maxAssetMb === 'number' && s.maxAssetMb > 0) ? Math.min(10, s.maxAssetMb) : 8,
    },
    denylist: {
      rules: Array.isArray(s.denyRules) ? clone(s.denyRules) : [],
      useDefaultProfile: s.denyDefaultProfile !== false,
    },
    analysis: {
      analyzeOnUpload: s.performAnalysisOnUpload === true,
      captureSourceMaps: s.captureSourceMaps !== false,
    },
  };
}

// Grouped effective config -> flat chrome.storage patch (what newSession writes to apply the
// resolved config to the live capture gate). Only sections present are mapped, so a partial
// effective (e.g. scope-only from an old build) leaves the other gate keys untouched.
// useDomainScope is derived (scope active iff any root domain), matching newSession's logic.
export function settingsFromConfig(effective) {
  const patch = {};
  for (const [section, leaf, storageKey] of SETTINGS_MAP) {
    const sec = effective && effective[section];
    if (sec && typeof sec === 'object' && leaf in sec) patch[storageKey] = clone(sec[leaf]);
  }
  if ('domainScopes' in patch) {
    patch.useDomainScope = Array.isArray(patch.domainScopes) && patch.domainScopes.length > 0;
  }
  return patch;
}
